Use XKCD API

drewpearce opened the pull request:
It uses the api instead of html scraping.

drewpearce committed:
1st Commit - It's operational

Uses the XKCD api (https://xkcd.com/json.html) to get the comics
instead of scraping the whole web page and regexing for the images and
alt text.
Needs some code cleanup and the random comic id should probably be it’s
own function.

drewpearce committed:
Functions are the best

Put the random comic id generation into a function

drewpearce committed:
Just call me Mr. Clean

cleaning up code to pass tests.

drewpearce committed:
Merge pull request #1 from drewpearce/use-xkcd-api

Use xkcd api

drewpearce committed:
added # nosec

Added # noses to rand usage so bandit passes. Rand is being used for
random selection from a list, nothing cryptographic.

bbriggs commented on legos/xkcd.py:
```diff
@@ -1,15 +1,20 @@
+            try:
+                return message['text'].split()[0] == '!xkcd'
+            except Exception as e:
+                logger.error('XKCD lego failed to check message text: %s' % e)
```
While we're already changing a bunch of stuff, let's convert this to the format() style interpolation

bbriggs commented on legos/xkcd.py:
```diff
@@ -1,15 +1,20 @@
+            try:
+                return message['text'].split()[0] == '!xkcd'
+            except Exception as e:
+                logger.error('XKCD lego failed to check message text: %s' % e)
```
We should also add a `raise()` here since that exception doesn't handle properly and will break the lego.

drewpearce replied:
saving for later

bbriggs commented on legos/xkcd.py:
```diff
@@ -57,21 +62,31 @@ def _build_url(self, comic_id):
+                comic_id = self._get_random_comic_id()
                 logger.debug('User requested comic by id: %s' % str(comic_id))
-                url = 'http://xkcd.com/%s' % str(comic_id)
+                url = 'http://xkcd.com/%s/info.0.json' % str(comic_id)
```
We should use the new interpolation style here too.

bbriggs commented on legos/xkcd.py:
```diff
@@ -57,21 +62,31 @@ def _build_url(self, comic_id):
-                url = 'http://dynamic.xkcd.com/random/comic'
-            else:
+                comic_id = self._get_random_comic_id()
                 logger.debug('User requested comic by id: %s' % str(comic_id))
```
Interpolation style

bbriggs replied:
Looks like you dropped the else statement here. The way this flowed was that if comic_id was `r` or `random`, then we'd fetch a random one. Otherwise, we'd log what ID was requested and pass that along.

bbriggs commented on legos/xkcd.py:
```diff
@@ -57,21 +62,31 @@ def _build_url(self, comic_id):
+            comic_id = random.randint(1, latest_json['num'])  # nosec
+        else:
+            logger.error('Requests encountered an error.')
+            logger.error('HTTP GET response code: %s' % latest.status_code)
```
Interpolation style

bbriggs commented on legos/xkcd.py:
```diff
@@ -57,21 +62,31 @@ def _build_url(self, comic_id):
+            logger.error('Requests encountered an error.')
+            logger.error('HTTP GET response code: %s' % latest.status_code)
+            latest.raise_for_status()
+            comic_id = 1337
```
Not a bad default ;)

bbriggs commented on legos/xkcd.py:
```diff
@@ -57,21 +62,31 @@ def _build_url(self, comic_id):
-        content = r.text
-        comic_regex = r'<div id="comic".*?\n?.*?(//im.+?)".+?\s?title="(.+?)"'
-        comic = re.search(comic_regex, content)
+        comic = json.loads(r.text)
```
I'm glad we got rid of that regex insanity.

bbriggs commented on legos/xkcd.py:
```diff
@@ -57,21 +62,31 @@ def _build_url(self, comic_id):
-            altText = comic.group(2).replace("&#39;", "'")
-            response = "%s %s" % (altText, "http:" + comic.group(1))
+            altText = comic['alt'].replace("&#39;", "'")
+            response = "%s %s" % (altText, comic['img'])
```
Interpolation style.

drewpearce committed:
PR Review changes

Switched to .format interpolation.
Also cleaned up mixed use of single vs double quotes.
Also added the comic # to the response.

drewpearce committed:
Merge branch 'master' into master
